### PR TITLE
CLR 003 - Login and Verify account

### PR DESCRIPTION
**Pull request by nguyenGALAXY**

### CODE REVIEW

File: src/auth/google.strategy.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import db from '../models/index'
+require('dotenv').config()
+passport.use(
+  new GoogleStrategy(
+    {
+      consumerKey:
+        '358036581199-d9odgr20b7jarvkv1falnsbq20cu2mt2.apps.googleusercontent.com',
+      consumerSecret: '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',
+      callbackURL: '/google_login/redirect',
+    },
```
Comment (nguyen315) [inline src/auth/google.strategy.js]: should we hide this inside .env file ?

File: src/controllers/auth.controller.js [hunk 2]
```diff
@@ -5,28 +5,112 @@ import { hashPassword } from '../utils/crypto'
...
+        { expiresIn: '5m' }
+      )
+      const url = `${process.env.FRONTEND_URL}/activateEmail/${activation_token}`
+      sendEmail(email, url)
+
+      res.status(httpStatusCodes.CREATED).json({
+        success: true,
+        message: 'Register Success! Please check your email to activate.',
+      })
+      //res.status(httpStatusCodes.CREATED).send(newUser)
```
Comment (nguyen315) [inline src/controllers/auth.controller.js]: remove comment code

File: src/controllers/auth.controller.js [hunk 3]
```diff
@@ -5,28 +5,112 @@ import { hashPassword } from '../utils/crypto'
...
+      })
+    }
     try {
       const newUser = await db.User.create({
         username,
         email,
         password: hash,
         firstName,
         lastName,
+        status: 'pending',
```
Comment (nguyen315) [inline src/controllers/auth.controller.js]: Can we create a constant.js file and export theses string from it
I think it is better to turn to upper case string for constant pending -> PENDING

File: src/controllers/auth.controller.js [hunk 4]
```diff
@@ -5,28 +5,112 @@ import { hashPassword } from '../utils/crypto'
...
+      const { email, status } = user.newUser
+      const checkEmail = await db.User.findOne({ where: { email: email } })
+      if (checkEmail) {
+        if (status === 'active') {
+          return res
+            .status(400)
+            .json({ success: true, message: 'Email already activated. ' })
+        }
+        const activeUser = await db.User.update(
+          { status: 'active' },
```
Comment (nguyen315) [inline src/controllers/auth.controller.js]: same as pending

File: src/controllers/auth.controller.js [hunk 5]
```diff
@@ -5,28 +5,112 @@ import { hashPassword } from '../utils/crypto'
...
+      res.status(httpStatusCodes.CREATED).json({
+        success: true,
+        message: 'Register Success! Please check your email to activate.',
+      })
+      //res.status(httpStatusCodes.CREATED).send(newUser)
     } catch (err) {
       res.status(httpStatusCodes.BAD_REQUEST).send(err.message)
     }
   }
+  @post('/activateEmail')
```
Comment (nguyen315) [inline src/controllers/auth.controller.js]: Change to /activate-email

File: src/controllers/auth.controller.js [hunk 6]
```diff
@@ -54,6 +141,53 @@ class AuthCtrl extends BaseCtrl {
       }
     )(req, res, next)
   }
-}
+  @post('/google_login')
```
Comment (nguyen315) [inline src/controllers/auth.controller.js]: Change to /google-login

File: src/controllers/auth.controller.js [hunk 7]
```diff
@@ -54,6 +141,53 @@ class AuthCtrl extends BaseCtrl {
...
+  async google_login(req, res) {
```
Comment (nguyen315) [inline src/controllers/auth.controller.js]: function name is camel so change to `googleLogin`

File: src/controllers/auth.controller.js [hunk 8]
```diff
@@ -54,6 +141,53 @@ class AuthCtrl extends BaseCtrl {
...
+      const { tokenId } = req.body
 
+      const account = await client.verifyIdToken({
+        idToken: tokenId,
+        audience: process.env.LOGIN_GOOGLE_CLIENT_ID,
+      })
+      const { email, name, family_name, given_name } = account.payload
+      const password = email + process.env.SECRET
+      const hash = await hashPassword(password)
+      const user = await db.User.authenticate(email, password)
```
Comment (nguyen315) [inline src/controllers/auth.controller.js:162]: What if user change their password? Then the authenticate always return `false`? We can discuss for better approach in this case

**Review by nguyen315 — COMMENTED** [review]
@nguyenGALAXY 1st round review

File: src/auth/google.strategy.js (hunk 1 above)
Comment (nguyenGALAXY) [reply src/auth/google.strategy.js]: i was delete this file

File: src/controllers/auth.controller.js (hunk 2 above)
Comment (nguyenGALAXY) [reply src/controllers/auth.controller.js]: already remove

File: src/controllers/auth.controller.js (hunk 3 above)
Comment (nguyenGALAXY) [reply src/controllers/auth.controller.js]: Yeah, I was created a constant.js file in folder utils

File: src/controllers/auth.controller.js (hunk 4 above)
Comment (nguyenGALAXY) [reply src/controllers/auth.controller.js]: already resolve

File: src/controllers/auth.controller.js (hunk 5 above)
Comment (nguyenGALAXY) [reply src/controllers/auth.controller.js]: yeah thanks

File: src/controllers/auth.controller.js (hunk 6 above)
Comment (nguyenGALAXY) [reply src/controllers/auth.controller.js]: ok bro

File: src/controllers/auth.controller.js (hunk 7 above)
Comment (nguyenGALAXY) [reply src/controllers/auth.controller.js]: sure

**Review by nguyen315 — COMMENTED** [review]
lgtm

**Pull request merged**